src/extension.ts

```diff
@@ -13,273 +13,412 @@
+        const workspace = new Workspace(folder);
+        workspaces.set(folder.uri.toString(), workspace);
+        workspace.start(context);
     }
-    try {
-        const childProcess = await childProcessPromise;
+}
 
-        childProcess.on('error', err => {
-            if ((<any>err).code == 'ENOENT') {
-                console.error('Could not spawn RLS process: ', err.message);
-                window.showWarningMessage('Could not start RLS');
-            } else {
-                throw err;
+let _sortedWorkspaceFolders: string[] | undefined;
+
+function sortedWorkspaceFolders(): string[] {
+    if (!_sortedWorkspaceFolders && workspace.workspaceFolders) {
+        _sortedWorkspaceFolders = workspace.workspaceFolders.map(folder => {
+            let result = folder.uri.toString();
+            if (result.charAt(result.length - 1) !== '/') {
+                result = result + '/';
             }
-        });
+            return result;
+        }).sort(
+            (a, b) => {
+                return a.length - b.length;
+            }
+        );
+    }
+    return _sortedWorkspaceFolders || [];
+}
 
-        if (CONFIGURATION.logToFile) {
-            const logPath = workspace.rootPath + '/rls' + Date.now() + '.log';
-            const logStream = fs.createWriteStream(logPath, { flags: 'w+' });
-            logStream.on('open', function (_f) {
-                childProcess.stderr.addListener('data', function (chunk) {
-                    logStream.write(chunk.toString());
-                });
-            }).on('error', function (err: any) {
-                console.error("Couldn't write to " + logPath + ' (' + err + ')');
-                logStream.end();
-            });
+function getOuterMostWorkspaceFolder(folder: WorkspaceFolder): WorkspaceFolder {
+    const sorted = sortedWorkspaceFolders();
+    for (const element of sorted) {
+        let uri = folder.uri.toString();
+        if (uri.charAt(uri.length - 1) !== '/') {
+            uri = uri + '/';
+        }
+        if (uri.startsWith(element)) {
+            return workspace.getWorkspaceFolder(Uri.parse(element)) || folder;
         }
-
-        return childProcess;
-    } catch (e) {
-        stopSpinner('RLS could not be started');
-        throw new Error('Error starting up rls.');
     }
+    return folder;
 }
 
-let lc: LanguageClient;
+function didChangeWorkspaceFolders(e: WorkspaceFoldersChangeEvent, context: ExtensionContext): void {
+    _sortedWorkspaceFolders = undefined;
 
-export async function activate(context: ExtensionContext) {
-    const promise = startLanguageClient(context);
-    configureLanguage(context);
-    registerCommands(context);
-    activateTaskProvider();
-    await promise;
+    for (let folder of e.added) {
+        folder = getOuterMostWorkspaceFolder(folder);
+        if (workspaces.has(folder.uri.toString())) {
+            continue;
+        }
+        for (const f of fs.readdirSync(folder.uri.fsPath)) {
```

Comment:
I don't think this works unfortunately, because at this point we're checking whether we *should* create a workspace and to use `RelativePattern` you need a workspace already created.